Add static_methods function

MasonProtter opened the pull request:
This is the result of some explorations I took in finding all the methods of a function at compile time, with back-edges built in. 
```julia
julia> using Tricks

julia> f(x) = x + 1
f (generic function with 1 method)

julia> static_methods(f)
# 1 method for generic function "f":
[1] f(x) in Main at REPL[2]:1

julia> @btime static_methods(f);
  43.324 ns (0 allocations: 0 bytes)

julia> @btime methods(f);
  2.250 μs (12 allocations: 720 bytes)
```
As you can see it happens at compile time, but also gets recompiled when new methods are added:
```julia
julia> f(::Int) = 1
f (generic function with 2 methods)

julia> static_methods(f)
# 2 methods for generic function "f":
[1] f(::Int64) in Main at REPL[6]:1
[2] f(x) in Main at REPL[2]:1

julia> f(x, y) = x / y
f (generic function with 3 methods)

julia> static_methods(f, Tuple{Int, Float64})
# 1 method for generic function "f":
[1] f(x, y) in Main at REPL[9]:1
```

I messed with the indentation to reflect regular module organization, but if you don't like that I can revert it.

oxinabox commented on src/Tricks.jl:
```diff
@@ -1,44 +1,89 @@
+_hasmethod_true(@nospecialize(f), @nospecialize(t)) = true
+
+"""
+            static_hasmethod(f, type_tuple::Type{<:Tuple)
```
need to unindent this also

oxinabox commented on src/Tricks.jl:
```diff
@@ -1,44 +1,89 @@
+
+static_methods(@nospecialize(f)) = _static_methods(Main, f, Tuple{Vararg{Any}})
+static_methods(@nospecialize(f) , @nospecialize(_T::Type)) = _static_methods(Main, f, _T)
+@generated function _static_methods(@nospecialize(m::Module), @nospecialize(f) , @nospecialize(_T::Type{T})) where {T <: Tuple}
```
What is the module doing?
Why is it set to `Main` and not say `Tricks` or the module that `f` comes from?

MasonProtter replied:
Right, I meant to ask about this and forgot. The module argument `m` get's passed to `expr_to_codeinfo` and then to `ccall(:jl_expand, Any, (Any, Any), ex, m)`. I'm not exactly sure what `jl_expand` does or what the proper default module would be. Perhaps @NHDaly knows? I cribbed `expr_to_codeinfo` from him at his suggestion.

oxinabox replied:
Anyway, don't need it see my suggestion in other comment

NHDaly replied:
Yeah i think it would be `Tricks`; i think it's the module where `jl_expand` should go looking for global references when it's expanding your function. (Not 💯% sure about that, but it's something like that.)

So yeah, you definitely don't need it to be a parameter here, you can just use `@__MODULE__` inside your `expr_to_codeinfo`.

But also i like @oxinabox's suggestion in his most recent comment, to just attach the edges to the codeinfo for exactly the function we want! It doesn't even need to be a generated function. In fact, you could probably just get the CodeInfo for `methods(f,t)` itself, attach the `edges` to it, and return that!

oxinabox replied:
> So yeah, you definitely don't need it to be a parameter here, you can just use @__MODULE__ inside your expr_to_codeinfo.

Lets make that change.

since we can't work out how to do this for generated functions just now.
(Though if one wants to go digging, seeing what Cassette/Zygote is doing might be insightful)

oxinabox commented on test/runtests.jl:
```diff
@@ -47,3 +47,11 @@ struct Foo end
+    f(x) = x + 1
+    @test (length ∘ collect ∘ static_methods)(f) == 1
+    f(::Int) = 1
+    @test (length ∘ collect ∘ static_methods)(f) == 2
```
Can you also check in the other order?

```
g(::Int) = 1
@test (length ∘ collect ∘ static_methods)(g) == 1
g(x) = x+1
@test (length ∘ collect ∘ static_methods)(g) == 2
```

MasonProtter replied:
Okay, so I added that to the tests and it fails. That's kinda spooky 'cause the other way works. Do you know why that is?

oxinabox replied:
I have a theory.  
The reason `f(::Int)` recompiles is because there is an edge to `f(::Any)` to the `static_methods`.
And the real purpose of that edge is designed to recompile normal functions.

For example: `foo(x::Int)=2f(x)` then if that was compiled and specialized on the fact that `x` was an `Int` then the static dispatch would be hard coded in to `f(::Any)` which at the time it was compiled was the clear correct method that would be called.
But then when `f(::Int)` is added then that is more specialized so need to know to recompile `foo` so it can see that.

But the reverse doesn't hold, if `foo` had compiled in a static dispatch to `g(::Int)` and you added `g(::Any)` then it doesn't need to recompile.
So it doesn't

> Something about the way I'm adding backedges isn't working.

Your backedges are fine.

--
Maybe we can just add the methd table one always, and never attach to methods.
idk if that will work

MasonProtter replied:
ahh, that makes sense. I tried always adding to the method table and never attaching to methods and it doesn't seem to work.

MasonProtter replied:
Wait, nvm it works!

oxinabox commented:
What if rather than dooing  all this stuff to create a `CodeInfo` for a function.
we just made a function that does what we want:
```
julia> @generated function _neomethods(f)
       methods(f.instance)
       end
_neomethods (generic function with 1 method)

julia> _neomethods(identity)
# 1 method for generic function "identity":
[1] identity(x) in Base at operators.jl:502

julia> @code_typed _neomethods(identity)
CodeInfo(
1 ─     return # 1 method for generic function "identity":
[1] identity(x) in Base at operators.jl:502
) => Base.MethodList

julia> Base.identity(::Float64) = 3

julia> _neomethods(identity)
# 1 method for generic function "identity":
[1] identity(x) in Base at operators.jl:502
```


So that, as you see does indeed compute the methods at compile time, but it is missing the hooks.

But then we do the same thing we do in `static_hasmethod`, and copy its code info
(we make it have the same signature as what ever outer method has)
and then attached hooks.

Probably extract out some common functionality between the two into helpers.

oxinabox commented on src/Tricks.jl:
```diff
@@ -1,44 +1,105 @@
+"""
+    static_hasmethod(f, type_tuple::Type{<:Tuple)
+
+Like `hasmethod` but runs at compile-time (and does not accept a worldage argument).
```
Doc string out of date

NHDaly commented:
So what happened with your investigation into whether the optimizer is already doing this?

oxinabox commented:
> So what happened with your investigation into whether the optimizer is already doing this?


The optimizer won't do *this*, it might do the thing that was Mason's original use case for this though

MasonProtter commented:
I'm not sure I understand exactly what you mean here

> But then we do the same thing we do in static_hasmethod, and copy its code info
(we make it have the same signature as what ever outer method has)
and then attached hooks.

or how to achieve it.

oxinabox commented:
Like in my example we define at the start :

```
@generated _neomethods(f, t) = methods(f,t)
```

Then in our `static_methods` we copy it's codeinfo, and add edges to it before returning it.

This we don't need to do the fiddly expr to codeinfo thing

NHDaly commented:
Yes, exactly!

Then your whole function would basically just be like:
```julia
@generated function static_methods(@nospecialize(m::Module), @nospecialize(f) , @nospecialize(_T::Type{T})) where {T <: Tuple}
        # Re-use the body of _neomethods:
        ci_orig = uncompressed_ast(typeof(_neomethods).name.mt.defs.func)
        ci = ccall(:jl_copy_code_info, Ref{CodeInfo}, (Any,), ci_orig)

        # And just add the edges so if a method is added to `f`, this recompiles
        mt = f.name.mt
        ci.edges = Core.Compiler.vect(mt, Tuple{Vararg{Any}})
        return ci
end
```

Or something else simple, more or less like that. :)

------

I'm actually not even sure if `_neomethods` needs to be `@generated` like you have it, @oxinabox

MasonProtter commented:
Unfortunately, if `_neomethods` is not `@generated`, then `static_methods` will run at runtime, not compiletime. 

If `_neomethods` is `@generated`, then `uncompressed_ast` doesn't work and julia suggests to use `code_lowered` instead, but when I switch to `code_lowered(_neomethods, Tuple{f, T})` I get
```
julia> begin
       using Tricks
       f(::Int) = 1
       static_methods(f, Tuple{Any})
       end
ERROR: Could not expand generator for `@generated` method MethodInstance for _neomethods(::typeof(f), ::Tuple{Any}). This can happen if the provided argument types (Tuple{typeof(f),Tuple{Any}}) are not leaf types, but the `generated` argument is `true`.
Stacktrace:
 [1] error(::String, ::Core.MethodInstance, ::String, ::String, ::Type, ::String, ::String) at ./error.jl:42
 [2] (::Base.var"#12#13"{Bool,DataType})(::Core.MethodInstance) at ./reflection.jl:806
 [3] _collect(::Array{Core.MethodInstance,1}, ::Base.Generator{Array{Core.MethodInstance,1},Base.var"#12#13"{Bool,DataType}}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at ./generator.jl:47
 [4] collect_similar(::Array{Core.MethodInstance,1}, ::Base.Generator{Array{Core.MethodInstance,1},Base.var"#12#13"{Bool,DataType}}) at ./array.jl:564
 [5] map(::Function, ::Array{Core.MethodInstance,1}) at ./abstractarray.jl:2073
 [6] #code_lowered#11 at ./reflection.jl:801 [inlined]
 [7] code_lowered at ./reflection.jl:793 [inlined]
 [8] #s15#7 at /Users/mason/.julia/dev/Tricks/src/Tricks.jl:69 [inlined]
 [9] #s15#7(::Any, ::Any, ::Any, ::Any) at none:0
 [10] (::Core.GeneratedFunctionStub)(::Any, ::Vararg{Any,N} where N) at ./boot.jl:524
 [11] top-level scope at REPL[1]:4
```

NHDaly commented:
> Unfortunately, if _neomethods is not @generated, then static_methods will
run at runtime, not compiletime.

Is that right? Even though `static_methods` itself is a @generated function?

On Sun, Jan 19, 2020 at 2:50 PM Mason Protter <notifications@github.com>
wrote:

> Unfortunately, if _neomethods is not @generated, then static_methods will
> run at runtime, not compiletime.
>
> If _neomethods is @generated, then uncompressed_ast doesn't work and
> julia suggests to use code_lowered instead, but when I switch to code_lowered(_neomethods,
> Tuple{f, T}) I get
>
> julia> begin
>        using Tricks
>        f(::Int) = 1
>        static_methods(f, Tuple{Any})
>        end
> ERROR: Could not expand generator for `@generated` method MethodInstance for _neomethods(::typeof(f), ::Tuple{Any}). This can happen if the provided argument types (Tuple{typeof(f),Tuple{Any}}) are not leaf types, but the `generated` argument is `true`.
> Stacktrace:
>  [1] error(::String, ::Core.MethodInstance, ::String, ::String, ::Type, ::String, ::String) at ./error.jl:42
>  [2] (::Base.var"#12#13"{Bool,DataType})(::Core.MethodInstance) at ./reflection.jl:806
>  [3] _collect(::Array{Core.MethodInstance,1}, ::Base.Generator{Array{Core.MethodInstance,1},Base.var"#12#13"{Bool,DataType}}, ::Base.EltypeUnknown, ::Base.HasShape{1}) at ./generator.jl:47
>  [4] collect_similar(::Array{Core.MethodInstance,1}, ::Base.Generator{Array{Core.MethodInstance,1},Base.var"#12#13"{Bool,DataType}}) at ./array.jl:564
>  [5] map(::Function, ::Array{Core.MethodInstance,1}) at ./abstractarray.jl:2073
>  [6] #code_lowered#11 at ./reflection.jl:801 [inlined]
>  [7] code_lowered at ./reflection.jl:793 [inlined]
>  [8] #s15#7 at /Users/mason/.julia/dev/Tricks/src/Tricks.jl:69 [inlined]
>  [9] #s15#7(::Any, ::Any, ::Any, ::Any) at none:0
>  [10] (::Core.GeneratedFunctionStub)(::Any, ::Vararg{Any,N} where N) at ./boot.jl:524
>  [11] top-level scope at REPL[1]:4
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oxinabox/Tricks.jl/pull/8?email_source=notifications&email_token=AAMCIELIYXSXUC3OTOMGSM3Q6SVINA5CNFSM4KIPDOJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJK2XSI#issuecomment-576039881>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMCIEMPXMXTPXNC5LSM5F3Q6SVINANCNFSM4KIPDOJQ>
> .
>

MasonProtter commented:
That’s what I saw testing it myself. It’s possible I did something incorrect though.

oxinabox commented:
hmm, I will have a play with this for a big see if i can make it work.

oxinabox commented:
I couldn't work it out either.
Lets make the change to get rid of the `::Module`  argument,
and update the readme,
and then we can merge it.

MasonProtter commented:
Okay, will do.

MasonProtter commented:
Okay, so I'd updated the docstring, switched to `@__MODULE__` in `expr_to_code_info` and added a test case where I make sure we're capturing the method-table for functions living in other modules. 

Tests pass locally.

Maybe we should get Travis set up for this package?

oxinabox commented on src/Tricks.jl:
```diff
@@ -1,44 +1,93 @@
+    ci = expr_to_codeinfo([Symbol("#self#"), :f, :_T], [:T], (:T,), :($ms))
+
+    method_insts = Core.Compiler.method_instances(f.instance, T, world)
+    method_doesnot_exist = isempty(method_insts)
```
```suggestion
```

oxinabox commented:
I have added Travis. If you rebase off master is should run.

Can you also bump the  version in the Project.toml so i can tag a release?
See:
https://white.ucc.asn.au/2019/09/28/Continuous-Delivery-For-Julia-Packages.html#what-if-i-dont-want-to-release-right-now--dev-versions

NHDaly commented on src/Tricks.jl:
```diff
@@ -1,44 +1,93 @@
+        Expr(Symbol("with-static-parameters"), lam, spnames...)
     end
+    ci = ccall(:jl_expand, Any, (Any, Any), ex, @__MODULE__)
+end
```
Can you actually rename this function to be less generic, since we've stripped some of its functionality, and i'd hat for people to try to reuse it, since it won't work.

I'd call this something like `create_codeinfo_with_returnvalue` or something like that, indicating that all this does is create a new CodeInfo that just returns the provided value.
And i'd also rename `e` to `value`, above.

(Also, I think you can probably get away with removing the inner-most `Expr(:block` from the labmda constructor above, since we're just returning `value` directly, but i'm not 100% on how it will affect the code generation)

MasonProtter replied:
Done

MasonProtter commented:
I think the CI failures are unrelated.

oxinabox commented:
urg, those test failures are unrelated yes.

I need to work out a better way to test these thins

oxinabox commented:
@MasonProtter  can you take a look at the commit I just mad and see that i didn't break anything?

I also note that since sparams is never `nothing` 
we could simplify to 
```
    expr = Expr(Symbol("with-static-parameters"),
        Expr(:lambda,
            argnames,
            Expr(Symbol("scope-block"),
                Expr(:block,
                    Expr(:return, value),
                )
            )
        )
        spnames...
    )
```

But I think that might make the code less clear

MasonProtter commented:
Your commit looks good to me and the tests pass locally.
